### ISSUE TITLE
[intersphinx] corrupt objects.inv due to zlib issue on windows

### ISSUE THREAD

**Issue by shimizukawa** [sphinx-doc/sphinx#524]
I am getting the following error when trying to use a generated objects.inv with intersphinx:

```
build succeeded, 1 warning.

WARNING: intersphinx inventory 'D:\\mrec3/sdspec/builtsrc/inmsx/html/objects.inv' not readable due to error: Error -3 while decompressing: invalid distance too far back
```

This error appears to be coming from zlib.

I would like to attach the offending file, and/or the project which generates it, but it is documentation for our internal proprietary C API for one of our core technologies at work.

I can not find an option to save the inventory in cleartext which would be an acceptable workaround for me.

---
- Bitbucket: https://bitbucket.org/birkenfeld/sphinx/issue/524
- Originally reported by: Anonymous
- Originally created at: 2010-09-14T09:33:07.321

**Comment by shimizukawa** [comment]
_From Anonymous on 2010-09-14 07:57:38+00:00_

After doing some more debugging, the issue appears to be with the chunk size being used. The inventory file I have is larger than the 16K (16*1024) chunk size used for reading in the compressed data. Now the decompressor should be properly managing it's internal buffer so the chunk size should not be an issue, but it is (python2.6). If I by hand read in the entire file into memory (after the text header) I can decompress everything in one go. But if it is chunked, it crashes.

I have implemented a temporary workaround by writing a version 2 to 1 inventory file converter, and converting all our object.inv file back to version 1 as part of our build system. This is.... well it sucks, but it does work.

I do not see any bugs about zlib having this issue on the python roundup tracker.

**Comment by shimizukawa** [comment]
_From Doug Napoleone on 2010-09-14 07:59:30+00:00_

Sorry, didn't realize I was not logged in. Adding a comment and tying this bug to my profile.

**Comment by shimizukawa** [comment]
_From Doug Napoleone on 2010-09-14 20:26:14+00:00_

**Found the bug!!!**

The problem is the inventory file is being opened in text mode, not binary mode:
{{{
#!python
def fetch_inventory(app, uri, inv):
    """Fetch, parse and return an intersphinx inventory file."""
    # both _uri_ (base URI of the links to generate) and _inv_ (actual
    # location of the inventory file) can be local or remote URIs
    localuri = uri.find('://') == -1
    join = localuri and path.join or posixpath.join
    try:
        if inv.find('://') != -1:
            f = urllib2.urlopen(inv)
        else:
            f = open(path.join(app.srcdir, inv))
}}}

That end 'open' should be "f = open(path.join(app.srcdir, inv), 'rb')"  The problem here is because the file is being opened in text mode, the binary data contains an EOF marker (windows in their infinite wisdom decided to have a character which can represent this so you can have data in a file after the EOF... yea windows!!!) So the read() later on truncates on this binary marker.

The problem here is that the header IS in text and will have a different EOL marker between windows and linux/mac. If you read in binary you need to deal with the potential for '\r' markers. I think this will also be a problem for URL based inventory files.

Note how the inventory file is written in the builders/html.py
{{{
#!python
    def dump_inventory(self):
        self.info(bold('dumping object inventory... '), nonl=True)
        f = open(path.join(self.outdir, INVENTORY_FILENAME), 'wb')
}}}

**Comment by shimizukawa** [comment]
_From Georg Brandl on 2010-09-15 15:53:57+00:00_

Argh, that's a shame.  I'll fix this ASAP and release a new 1.0 bugfix version.

**Comment by shimizukawa** [comment]
_From Georg Brandl on 2010-09-15 15:55:35+00:00_

Ah yes, can you do me a favor and verify that URLopened files do not exhibit that behavior?

**Comment by shimizukawa** [comment]
_From Doug Napoleone on 2010-09-15 15:58:16+00:00_

After looking things over just adding the 'rb' should do the trick and the urlopened files should be fine. I can test the urlopen issue, but I might not get to it today. I will try. Thanks for the fast response!

**Comment by shimizukawa** [comment]
_From Doug Napoleone on 2010-09-16 19:06:26+00:00_

Sorry it took so long to do this simple test.
I tested the objects.inv which was causing me problems on windows using apache servers running on both windows and REL, from remote machines running win64, OSX, and REL. There were no problems decompressing the data, and urlopen is the equivalent of a read-binary.

**Comment by shimizukawa** [comment]
_From Georg Brandl on 2010-09-17 07:58:31+00:00_

OK, nice.  Fixed in <<cset 81bb0cd8d1ba162e5093ee7689bc947f9bdecdf4>>.

**Comment by shimizukawa** [comment]
_From Georg Brandl on 2014-10-18 07:30:10+00:00_

Removing version: 1.0.3 (automated comment)
